icu4c/source/common/rbbiscan.cpp

```diff
@@ -530,6 +530,11 @@ UBool RBBIRuleScanner::doParseActions(int32_t action)
             n = fNodeStack[fNodeStackPtr];
             uint32_t v = u_charDigitValue(fC.fChar);
             U_ASSERT(v < 10);
+            // Avoid overflow n->fVal
+            if (n->fVal > static_cast<int32_t>((INT32_MAX - v) / 10)) {
```

Comment:
It might be clearer to compute the value in a temporary of type int64_t. The 32 bit overflow check inside the loop is then much simpler
```temp > INT32_MAX```